src/Functions/array/arrayMannWhitneyUTest.h

```diff
@@ -0,0 +1,201 @@
+#include <DataTypes/DataTypesNumber.h>
+#include <Common/HashTable/ClearableHashMap.h>
+#include <Common/ColumnsHashing.h>
+#include <Functions/array/arrayStatTest.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+}
+
+enum class Alternative {
+    TwoSided,
+    Greater,
+    Less
+};
+
+
+/** Calculates Mann-Whitney U nonparametric statistical test of the null hypothesis that it is
+  * equally likely that a randomly selected value from one population will be less than or greater
+  * than a randomly selected value from a second population. This test can be used to investigate
+  * whether two independent samples were selected from populations having the same distribution
+  *
+  * The function takes two arrays of samples
+  * These values are considered as the control and test samples of some observation
+  *
+  * https://en.wikipedia.org/wiki/Mann–Whitney_U_test
+  *
+  * Function returns p-value assuming an asymptotic normal distribution.
+  */
+
+
+template <typename Name, Alternative alternative>
+class ArrayMannWhitneyUTestImpl
+{
+private:
+    /// Initially allocate a piece of memory for 512 elements. NOTE: This is just a guess.
+    static constexpr size_t INITIAL_SIZE_DEGREE = 9;
+
+public:
+    static constexpr auto name = Name::name;
+    using ResultType = Float64;
+
+    static String getName()
+    {
+        return name;
+    }
+
+    static DataTypePtr getReturnType(const DataTypePtr &v1, const DataTypePtr &v2)
+    {
+        if (!(isNumber(v1) && isNumber(v2)))
+            throw Exception(std::string(getName()) + " label must have numeric type.", ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        return std::make_shared<DataTypeNumber<ResultType>>();
+    }
+
+
+    template <class ForwardIt, class T>
+    static void iota(
+        ForwardIt first,
+        ForwardIt last,
+        T value)
+    {
+        while(first != last)
+        {
+            *first++ = value;
+            ++value;
```

Comment:
Why do you try to write `++first` and `*first` in one row, but value is incremented on a separate line?
If it is not for optimization purpose it is better to split into trivial lines.